### PR TITLE
Unit Tests added for dropdown

### PR DESCRIPTION
**Pull request by alliclemens**
Partial test suite added for dropdown component

Also added scrollbar to browser test results to allow viewing of all test results

### CODE REVIEW

**Comment by DanielRuf** [comment]
Awesome, thanks for the PR. I'll take a closer look at it and test it on the terminal and in the browser.

**Comment by alliclemens** [comment]
@DanielRuf would it be alright if I actually closed this pull request and moved the changes regarding the scrollbar into a different PR?
 I realized after submitting that would probably be better practice, and I also have some other comments relating to overall testing and modifications to the wiki documentation that would go along with it.

**Comment by DanielRuf** [comment]
> @DanielRuf would it be alright if I actually closed this pull request and moved the changes regarding the scrollbar into a different PR?

You can try `git revert 2d56de0b675b41d2205d051651989ea98b672dfa` and then in another branch use `git cherrypick 2d56de0b675b41d2205d051651989ea98b672dfa`

https://stackoverflow.com/a/2474430/753676

File: test/javascript/components/dropdown.js [hunk 1]
```diff
@@ -19,6 +25,7 @@ describe('Dropdown', function() {
       $dropdownController = $(getDropdownController()).appendTo('body');
       $dropdownContainer = $(getDropdownContainer()).appendTo('body');
       plugin = new Foundation.Dropdown($dropdownContainer, {});
+      console.log('dropdown constructor');
```
Comment (DanielRuf) [inline test/javascript/components/dropdown.js]: We already output `constructor()`. I guess this line here is not needed.

File: test/javascript/components/dropdown.js [hunk 2]
```diff
@@ -37,6 +44,54 @@ describe('Dropdown', function() {
...
+      plugin.open();
+
+      $dropdownController.on('show.zf.dropdown', function() {
+        $('#my-dropdown').should.not.be.hidden;
+        $('#my-dropdown').should.have.attr('aria-hidden', 'false');
+        $('#my-dropdown').should.have.class('is-open');
+        done();
+      });
+      plugin.close();
+    });
```
Comment (ncoden) [inline test/javascript/components/dropdown.js]: These functions do not test "click" and "hover" dropdowns but if the `.open()` works.

So you can delete one of them and replace with simply `should open dropdown`, OR provide a test that simulate the `hover` to test if it opens the dropdown (and close after).

File: test/javascript/components/dropdown.js [hunk 3]
```diff
@@ -37,6 +44,54 @@ describe('Dropdown', function() {
...
+      $dropdownContainer = $(getAutoFocusDropdownContainer()).appendTo('body');
+      plugin = new Foundation.Dropdown($dropdownContainer, {});
+      plugin.open();
+
+      $focusedElement = '<input type="text" placeholder="should auto focus" id="focus-text" />';
+
+      $dropdownController.on('show.zf.dropdown', function() {
+        $('#my-dropdown').should.not.be.hidden;
+        $('#my-dropdown').should.have.attr('aria-hidden', 'false');
+        $('#my-dropdown').should.have.class('is-open');
```
Comment (ncoden) [inline test/javascript/components/dropdown.js]: the 3 checks above are not relevant here as we only want to test if the element is focused.

File: test/javascript/components/dropdown.js (hunk 1 above)
Comment (ncoden) [inline test/javascript/components/dropdown.js]: You let a `console.log` behind ;)

File: test/javascript/components/dropdown.js [hunk 4]
```diff
@@ -37,6 +44,54 @@ describe('Dropdown', function() {
...
+      plugin.close();
+    });
+
+    it('should autofocus input on open', function() {
+      $dropdownController = $(getDropdownController()).appendTo('body');
+      $dropdownContainer = $(getAutoFocusDropdownContainer()).appendTo('body');
+      plugin = new Foundation.Dropdown($dropdownContainer, {});
+      plugin.open();
+
+      $focusedElement = '<input type="text" placeholder="should auto focus" id="focus-text" />';
```
Comment (ncoden) [inline test/javascript/components/dropdown.js]: We can also put an ID on the autofocus (that we pass as parameter) and test it on the `activeElement`. I find that more enduring.

File: test/javascript/index.html [hunk 5]
```diff
@@ -7,11 +7,12 @@
     <script src="../../node_modules/jquery/dist/jquery.js"></script>
     <script src="../../_build/assets/js/foundation.js"></script>
     <script src="lib/mocha.js"></script>
+    <link rel="stylesheet" href="lib/custom-mocha.css" />
     <link rel="stylesheet" href="lib/mocha.css" />
     <link rel="stylesheet" href="../../_build/assets/css/foundation-float.css">
     <link rel="stylesheet" href="../../node_modules/motion-ui/dist/motion-ui.css">
   </head>
-  <body>
+  <body id="test-body">
```
Comment (ncoden) [inline test/javascript/index.html]: I don't understand why this is in this PR. But as you said it, you can move this change to its own PR.

**Review by ncoden — CHANGES_REQUESTED** [review]
Hi @alliclemens !

Thank you a lot. It's not common at all to receive unit tests in a first contribution. 👏

File: test/javascript/components/dropdown.js [hunk 6]
```diff
@@ -37,6 +44,54 @@ describe('Dropdown', function() {
...
       Foundation.Keyboard.trapFocus.restore();
     });
+
+    it('should open dropdown on button click', function() {
+      $dropdownController = $(getDropdownController()).appendTo('body');
+      $dropdownContainer = $(getDropdownContainer()).appendTo('body');
+      plugin = new Foundation.Dropdown($dropdownContainer, {});
+      plugin.open();
+
+      $dropdownController.on('show.zf.dropdown', function() {
```
Comment (DanielRuf) [inline test/javascript/components/dropdown.js]: The done callback is missing in the parameters.

File: test/javascript/components/dropdown.js [hunk 7]
```diff
@@ -37,6 +44,54 @@ describe('Dropdown', function() {
...
+      plugin.close();
+    });
+
+    it('should open dropdown on hover', function() {
+      $dropdownController = $(getDropdownController()).appendTo('body');
+      $dropdownContainer = $(getHoverDropdownContainer()).appendTo('body');
+      plugin = new Foundation.Dropdown($dropdownContainer, {});
+      plugin.open();
+
+      $dropdownController.on('show.zf.dropdown', function() {
```
Comment (DanielRuf) [inline test/javascript/components/dropdown.js]: The done callback is missing in the parameters.

File: test/javascript/components/dropdown.js [hunk 8]
```diff
@@ -37,6 +44,54 @@ describe('Dropdown', function() {
       sinon.assert.called(spy);
       Foundation.Keyboard.trapFocus.restore();
     });
+
+    it('should open dropdown on button click', function() {
```
Comment (DanielRuf) [inline test/javascript/components/dropdown.js]: The done callback is missing in the parameters.

**Review by DanielRuf — COMMENTED** [review]
I've done some testing.

Thanks for the unit tests, they are useful.

The code should be the following:

```
    it('should open dropdown on button click', function (done) {
      $dropdownController = $(getDropdownController()).appendTo('body');
      $dropdownContainer = $(getDropdownContainer()).appendTo('body');
      plugin = new Foundation.Dropdown($dropdownContainer, {});

      plugin.$element.on('show.zf.dropdown', function () {
        plugin.$element.should.be.visible;
        plugin.$element.should.have.attr('aria-hidden', 'false');
        plugin.$element.should.have.class('is-open');
        done();
      });
      $dropdownController.trigger("click");
    });

    it('should open dropdown on hover', function (done) {
      $dropdownController = $(getDropdownController()).appendTo('body');
      $dropdownContainer = $(getHoverDropdownContainer()).appendTo('body');
      plugin = new Foundation.Dropdown($dropdownContainer, {});

      plugin.$element.on('show.zf.dropdown', function () {
        plugin.$element.should.be.visible;
        plugin.$element.should.have.attr('aria-hidden', 'false');
        plugin.$element.should.have.class('is-open');
        done();
      });
      $dropdownController.trigger("mouseenter");
    });

    it('should autofocus input on open', function (done) {
      $dropdownController = $(getDropdownController()).appendTo('body');
      $dropdownContainer = $(getAutoFocusDropdownContainer()).appendTo('body');
      plugin = new Foundation.Dropdown($dropdownContainer, {});

      plugin.$element.on('show.zf.dropdown', function () {
        plugin.$element.should.be.visible;
        plugin.$element.should.have.attr('aria-hidden', 'false');
        plugin.$element.should.have.class('is-open');
        document.activeElement.should.be.equal($dropdownContainer.find('#focus-text').get(0));
        done();
      });
      plugin.open();
    });
```

Can you update the PR to reflect this?
Let me know if I should explain why it should be like this and what was wrong if it is not clear.

File: test/javascript/components/dropdown.js (hunk 3 above)
Comment (DanielRuf) [reply test/javascript/components/dropdown.js]: Right. I guess we can safely remove them from the one unit test.

**Comment by DanielRuf** [comment]
> Thank you a lot. It's not common at all to receive unit tests in a first contribution.

Definitely awesome for the first contribution =)

**Comment by ncoden** [comment]
@DanielRuf See my review just above.

**Comment by ncoden** [comment]
@alliclemens Will you have some time to take care of the requested changes ? Would you need any help with it ?

**Comment by ncoden** [comment]
Hi @alliclemens. I made some changes.

> * split tests about `open()` and mouse events
> * tests triggered events instead of method calling for mouse events
> * add missing `done()` when test is asynchronous
> * rely on options passed to plugin instead of in the HTML
> * test the Id of the focused element instead of the whole HTML (it can change)
> * improve various tests names

Could you say what you think of it ?

Also, I don't understand `overflow: scroll;` added on body. I have no problem with scrolling on the test page (macOS Chrome `64.0.3282.186`). Are you sure it is required ?

Poke @DanielRuf for review ;)

**Comment by DanielRuf** [comment]
> Also, I don't understand overflow: scroll; added on body. I have no problem with scrolling on the test page (macOS Chrome 64.0.3282.186). Are you sure it is required ?

It should be `auto`.

Without this you can not scroll the results of mocha in the browser.

**Comment by DanielRuf** [comment]
You can try it when you click on one of the test failures. I'm sure there is a scrolling issue.

**Comment by ncoden** [comment]
> Without this you can not scroll the results of mocha in the browser.
> You can try it when you click on one of the test failures. I'm sure there is a scrolling issue.

I can't reproduce this. On which browser are you testing ?

**Comment by DanielRuf** [comment]
> I can't reproduce this. On which browser are you testing ?

Also Chrome. Will write down the steps to reproduce this in about 8 hours if this is ok.

**Comment by ncoden** [comment]
No problem. Good night 😄

**Comment by DanielRuf** [comment]
> No problem. Good night 😄

Sorry for the late reply.

It seems the manual `overflow` setting is not needed anymore. I recall that there was an issue with there scrolling but I can not reproduce it anymore with the current develop branch.

Was is due to some uncaught error / exception? Mocha manually applies an `overflow` setting when the testsuite is run (adds and removes it using JS).

So currently we do not need the `overflow`, right

**Comment by ncoden** [comment]
I don't see any `overflow` property or class/ids applied when the JS is running. I revert the commit as I have no trouble to scroll in the tests.

@alliclemens Please come back to us with more more informations if we are wrong.

**Comment by DanielRuf** [comment]
I guess this is safe now for a merge or do we still need something in this PR?

**Review by ncoden — APPROVED** [review]
@DanielRuf No we're good
